src/runtime-rs/crates/resource/src/network/utils/link/manager.rs

```diff
@@ -181,119 +181,40 @@ fn parse_bridge(mut ibs: Vec<InfoBridge>) -> Bridge {
     }
     bridge
 }
-#[derive(Debug, PartialEq, Eq, Clone, Default)]
-pub struct Device {
-    attrs: Option<LinkAttrs>,
-}
-
-impl Link for Device {
-    fn attrs(&self) -> &LinkAttrs {
-        self.attrs.as_ref().unwrap()
-    }
-    fn set_attrs(&mut self, attr: LinkAttrs) {
-        self.attrs = Some(attr);
-    }
-    fn r#type(&self) -> &'static str {
-        "device"
-    }
-}
-
-#[derive(Debug, PartialEq, Eq, Clone, Default)]
-pub struct Tuntap {
-    pub attrs: Option<LinkAttrs>,
-}
-
-impl Link for Tuntap {
-    fn attrs(&self) -> &LinkAttrs {
-        self.attrs.as_ref().unwrap()
-    }
-    fn set_attrs(&mut self, attr: LinkAttrs) {
-        self.attrs = Some(attr);
-    }
-    fn r#type(&self) -> &'static str {
-        "tuntap"
-    }
-}
-
-#[derive(Debug, PartialEq, Eq, Clone, Default)]
-pub struct Veth {
-    attrs: Option<LinkAttrs>,
-
-    /// on create only
-    pub peer_name: String,
-}
-
-impl Link for Veth {
-    fn attrs(&self) -> &LinkAttrs {
-        self.attrs.as_ref().unwrap()
-    }
-    fn set_attrs(&mut self, attr: LinkAttrs) {
-        self.attrs = Some(attr);
-    }
-    fn r#type(&self) -> &'static str {
-        "veth"
-    }
-}
-
-#[derive(Debug, PartialEq, Eq, Clone, Default)]
-pub struct IpVlan {
-    attrs: Option<LinkAttrs>,
-
-    /// on create only
-    pub peer_name: String,
-}
 
-impl Link for IpVlan {
-    fn attrs(&self) -> &LinkAttrs {
-        self.attrs.as_ref().unwrap()
-    }
-    fn set_attrs(&mut self, attr: LinkAttrs) {
-        self.attrs = Some(attr);
-    }
-    fn r#type(&self) -> &'static str {
-        "ipvlan"
-    }
-}
-
-#[derive(Debug, PartialEq, Eq, Clone, Default)]
-pub struct MacVlan {
-    attrs: Option<LinkAttrs>,
-
-    /// on create only
-    pub peer_name: String,
-}
-
-impl Link for MacVlan {
-    fn attrs(&self) -> &LinkAttrs {
-        self.attrs.as_ref().unwrap()
-    }
-    fn set_attrs(&mut self, attr: LinkAttrs) {
-        self.attrs = Some(attr)
-    }
-    fn r#type(&self) -> &'static str {
-        "macvlan"
-    }
+macro_rules! impl_network_dev {
+    ($r_type: literal , $r_struct: ty) => {
+        impl Link for $r_struct {
+            fn attrs(&self) -> &LinkAttrs {
+                self.attrs.as_ref().unwrap()
+            }
+            fn set_attrs(&mut self, attr: LinkAttrs) {
+                self.attrs = Some(attr);
+            }
+            fn r#type(&self) -> &'static str {
+                $r_type
+            }
+        }
+    };
 }
 
-#[derive(Debug, PartialEq, Eq, Clone, Default)]
-pub struct Vlan {
-    attrs: Option<LinkAttrs>,
+macro_rules! define_and_impl_network_dev {
+    ($r_type: literal , $r_struct: tt) => {
+        #[derive(Debug, PartialEq, Eq, Clone, Default)]
+        pub struct $r_struct {
+            attrs: Option<LinkAttrs>,
+        }
```

Comment:
What about 
```
    /// on create only
    pub peer_name: String,
```
in `pub struct` definition
for `MacVlan`,  `IpVlan` and `Veth`.

Reply:
I didn't find any reference for this field, is it ready for feature usage?

Reply:
yup, we can always add it back if needed.

Reply:
I have checked `peer_name` and it is not used now.
We could remove it now and add it back if needed.